_posts/2020-05-26-introduction-to-spring-webflux.md

```diff
@@ -0,0 +1,170 @@
+
+### Tips for some common pitfalls you might encounter
+
+#### Returning an empty value
+Sometimes you just need to check something in a function, without returning a value. For example, you might need a function to check if some operation can be made. Your function returns a void value in a success case and throws an exception in an error case:
+```java
+public Foo doSomething() {
+
+  // Get state etc.
+  Bar bar = getBar();
+
+  checkOperationXxx(bar);
+
+  return operationXxx(bar); // Returns foo
+}
+
+void checkOperationXxx(Bar bar) {
+  // Check for access rights, etc.
+  if (bar.isOperationXxxAllowed()) {
+    return;
+  }
+  throw new OperationDeniedException();
+}
+```
+You might try to implement this with WebFlux as follows:
+```java
+public Mono<Foo> doSomething() {
+  return getBar() // Returns Mono<Bar>
+    .map(bar -> checkOperationXxx(bar)
+      .onErrorResume(throwable -> { // Handle error })
+      .map(aValue -> operationXxx(bar)));
+}
+
+Mono<Void> checkOperationXxx(Bar bar) {
+  // Check for access rights, etc.
+  if (bar.isOperationXxxAllowed()) {
+    return Mono.empty();
+  }
+  return Mono.error(new OperationDeniedException());
+}
+```
+However, when you return an empty value with `Mono.empty()` the result really is an empty value. You can't execute any operation such as map or flatMap to it. Your `operationXxx()` is never executed and you'll spend hours trying to figure out what's wrong.
+
+Your next attempt might be to write the caller function as:
+```java
+public Mono<Foo> doSomething() {
+  return getBar() // Returns Mono<Bar>
+    .map(bar -> checkOperationXxx(bar)
+      .onErrorResume(throwable -> { // Handle error })
+      .then(operationXxx(bar)));
+}
+```
+This does work in an success case: `checkOperationXxx()` completes with empty value and then you start executing a new Mono in `then()`-function. However, `operationXxx()` is executed also in an error case! This is not what you want.
+
+In our case we solved the issue by always returning a non-empty value, usually the same value which we pass to the function as a parameter:
+```java
+public Mono<Foo> doSomething() {
+  return getBar() // Returns Mono<Bar>
+    .map(bar -> checkOperationXxx(bar)
+      .onErrorResume(throwable -> { // Handle error })
+      .map(b -> operationXxx(b)));
+}
+
+Mono<Bar> checkOperationXxx(Bar bar) {
+  // Check for access rights, etc.
+  if (bar.isOperationXxxAllowed()) {
+    return Mono.just(bar);
+  }
+  return Mono.error(new OperationDeniedException());
+}
+```
+
+This works as you want - `checkOperationXxx()` does the necessary checks and `operationXxx()` is executed only on a success case.
```

Comment:
'This works as you expect:  `checkOperationXxx()` does the necessary checks and `operationXxx()` is executed only in case of a success.'